JavaScript/Angular.md

```diff
@@ -0,0 +1,687 @@
+
+### Recommended Editor
+
+Recommended editor for Angular development is [VSCode](https://code.visualstudio.com/) as it has great integration with TypeScript and has some handy Angular plugins (e.g. [Angular Language Service](https://marketplace.visualstudio.com/items?itemName=Angular.ng-template)). This is just a recommendation, please feel free to use whatever editor you are most comfortable with.
+
+## File/Module Organization and Naming
+
+If you created the app using Angular CLI, there is already some clear structure in place. Building on that structure, you will probably want to add some more dirs to help you organize your code.
+
+Try to split up your application into components which you can reuse. If, for example, you have to handle lazy loading of all images in the app, create a `Image` module and component in `app/components` and use it in other components whenever you have to show an image. Export the declared component and import/provide dependencies which your component has.
+
+Container components which are loaded directly via routing (by either `component` or `loadChildren`) should be placed in `app/pages` dir. If component is loaded because you have used it in a template, that component should be in `app/components`
+
+One little exception here: if there is some very specific component tied to some routing submodule which is used only in that submodule, it is OK to leave it in that submodule (but as soon as it is used in another routing submodule, move it to `app/components`.
+
+Store all your animations in `app/animations/` and import them for usage in component decorators.
+
+Classes which might be used throughout the application (e.g. generic `Dictionary` class) should be in `app/classes`.
+
+Entity models (`User`, `Post`, etc.) should be in `app/models`.
+
+If you are using some JS libs which do not have typings, add your own custom typing in `app/custom-typings`.
+
+Whenever you have some attribute whose value can be some value from a set of values, make an enum and store that enum in `app/enums`. Good example is an enum for HTTP status codes. If you need some metadata for the enum, create an enum data file which exportst what is basically a dictionary with enum values used as keys and use whatever you need for values. One such example would be: `export const httpStatusCodeData = { [HttpStatus.OK_200]: { translation: 'Success' } }`
+
+If you are using `ngx-form-object`, place `FormObject`s, `FromStore`s and validators in `app/forms`. If you have form components which are reused throughout the app, put them in `app/components/forms`.
+
+Put all route guards in `app/guards`.
+
+If you need some helper functions (like a rounding function which actually rounds correctly *cough* Math.round *cough* floating point *cough*), put the helper in `app/helpers` in its own dir, make sure to write tests for it and simply export it in `app/helpers/helpers.ts`.
+
+Save all your interface files to `app/interfaces`.
```

Comment:
Maybe mention the naming preference of prefixing the names with the capital *I* (`IUser` instead of `User`)

Reply:
I think that's up for debate, as it doesn't exactly abide with Angular style guide https://angular.io/guide/styleguide#interfaces

Reply:
My main argument for prefixing interfaces with `I` are cases like this: https://github.com/infinum/frontend-handbook/blob/4a864ae22df273ef8f7d4d5b5b87a1def549720d/JavaScript/Angular.md#testing-a-service-example

Reply:
From the angular docs:

```
Consider naming an interface without an I prefix.
...
TypeScript guidelines discourage the I prefix.
```

@fvoska I don't get your example.

Reply:
There is a service, service implements a interface and testing/stub service implements the same interface as well

Reply:
When you open the [TypeScript guidelines](https://github.com/Microsoft/TypeScript/wiki/Coding-guidelines) link from the Angular Styleguide you see this:

>**STOP READING IMMEDIATELY**
>**THIS PAGE PROBABLY DOES NOT PERTAIN TO YOU**
>**These are Coding Guidelines for Contributors to TypeScript**
>**This is NOT a prescriptive guideline for the TypeScript community**

😅

Reply:
> There is a service, service implements a interface and testing/stub service implements the same interface as well

And what's wrong with that?

ie.
```
// service/user/user.service.interface.ts
export interface UserService {
  user$: Observable<UserModel>;
}

// service/user/user.service.ts
@Injectable({ providedIn: 'root' })
export class UserService implements UserService {
  public user$: Observable<UserModel>;
  //...
}

// service/user/user.testing.service.ts
export class UserTestingService implements UserService {
  public user$: Observable<UserModel> = of(new UserModelStub('steve'));
}
```

Reply:
you can not write `UserService implements UserService`, you either have to rename during import or export under a different name.

Reply:
> export class UserService implements UserService
Wouldn't this be a bit confusing?

(reminds me of `@Inject(SOMETHING) private something: Something` 😆 )

Reply:
🤦‍♂️  you are correct.
In that case you will have to use named import for UserService interface.
But you will use that in a single file so I'm still against prefixing.
What others think?

Reply:
What we could do as well is remove interface altogether do something like this:

```ts
// service/user/user.service.ts
@Injectable({ providedIn: 'root' })
export class UserService implements UserService {
  public user$: Observable<UserModel>;
  //...
}

// service/user/user.testing.service.ts
export class UserTestingService implements Partial<UserService> {
  public user$: Observable<UserModel> = of(new UserModelStub('steve'));
}
```

But in this approach there is a problem - you are not enforced to implement anything in testing service.

Reply:
Coming from a Java background, prefixing with `I` is natural for me.

Reply:
Why do you need `Partial`?

> Coming from a Java background, prefixing with `I` feels natural to me.
I know, but most of the Angular projects/libraries won't/don't have prefixed interfaces.

Reply:
`export class UserService implements UserService {`
This is too confusing... Considering it's also possible to implement classes, It might lead to issues.